### ISSUE TITLE
Example fails - due to usage of dot in bucket name

### ISSUE THREAD

**Issue by ketanm** [diagnosia/flutter_aws_s3_client#4]
Hi,

In the example code [main.dart](https://github.com/diagnosia/flutter_aws_s3_client/blob/master/example/lib/main.dart) AwsS3Client is initialized with bucket argument as `bucketid.s3.$region.amazonaws.com`? Assuming that `bucketid` is the actual name of the bucket, e.g. 'mybucket'. This does not appear to work. The implementation of `AwsS3Client` also makes the assumption of s3 via `_service` variable. 

Please can you clarify what is the correct usage?

Ketan

**Comment by knaeckeKami** [comment]
Hi,

You're right, the naming is not ideal and the example code could be improved.
The correct initialisation of the client is like this:

```dart
const region = "eu-central-1";
const bucketId = "myBucket";

final client = AwsS3Client(
        region: region,
        bucket: "$bucketId.s3.$region.amazonaws.com",
        accessKey: <your access key>,
        secretKey: <your secret key>);
```

Can you clarify what kind of problems you run into?
Another user had issues with incorrect config ( #3 ), and the problem was obfuscated by weird behaviour of the underlying http client (from the http package https://pub.dev/packages/http ).

Maybe use the `buildSignedGetParams` method, inspect the results (url and headers) and try to do the request with another http client (e.g. `HttpClient`, an example of how to use it is at the bottom of the README)

**Comment by ketanm** [comment]
Hi @knaeckeKami 

Thanks for the quick response.

I used as shown above but it kept giving me Handshake error. I will try your suggestion and let you know.

Ketan

**Comment by ketanm** [comment]
This is the exact error:
```
final AwsS3Client s3client = AwsS3Client(
      region: REGION,
      bucket: "$BUCKET.s3.$REGION.amazonaws.com",
      accessKey: ACSSES_KEY,
      secretKey: SECRET_KEY);

  // SignedRequestParams signedPram = s3client.buildSignedGetParams(path: "stream/MIL-MOLNG");
  
  var listBucketResult =
      await s3client.listObjects(prefix: "stream/MIL-MOLNG/", delimiter: "/");
```

Exception trace is as below:
```
2019-11-22 15:17:28.078030-0500 Runner[348:8127] [VERBOSE-2:ui_dart_state.cc(148)] Unhandled Exception: HandshakeException: Handshake error in client (OS Error: 
	CERTIFICATE_VERIFY_FAILED: ok(handshake.cc:352))
#0      IOClient.send (package:http/src/io_client.dart:33:23)
<asynchronous suspension>
#1      BaseClient._sendUnstreamed (package:http/src/base_client.dart:169:38)
<asynchronous suspension>
#2      BaseClient.get (package:http/src/base_client.dart:32:7)
#3      get.<anonymous closure> (package:http/http.dart:46:36)
#4      _withClient (package:http/http.dart:166:20)
<asynchronous suspension>
#5      get (package:http/http.dart:46:5)
#6      AwsS3Client._doSignedGetRequest (package:flutter_aws_s3_client/src/client/client.dart:99:12)
<asynchronous suspension>
#7      AwsS3Client.listObjects (package:flutter_aws_s3_client/src/client/client.dart:32:28)
<asynchronous suspension>
#8      main (package:example/main.dart:20:22)
<asynchronous suspension>
#9      _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:229:25)
#10     _rootRun (dart:async/zone.dart:1124:13)
#11     _CustomZone.run (dart:async/zone.dart:1021:19)
#12     _runZoned (dart:async/zone.dart:1516:10)
#13     runZoned (dart:async/zone.dart:1500:12)
#14     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:221:5)
#15     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:305:19)
#16     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)
```

I will look into it as to why it is failing.

**Comment by knaeckeKami** [comment]
Interesting. 
Do you use an http proxy (often required in many company or university networks)?

Does the http client work with other urls?
e.g.:   

```dart
import 'package:http/http.dart';

void main() async {

final result =await Client().get("https://google.com");
debugPrint(result.body);
}
```

**Comment by ketanm** [comment]
Works as expected

```flutter: <!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="en"><head><meta content="Search the world's information, including webpages, images, videos and more. Google has many special features to help```

I noticed that this package is using `https://bucket_id.s3.region.amazonaws.com/?list-type...` instead of `https://s3.region.amazonaws.com/bucket_id?list-type...`. I think both are supported but wondering if there would be any issue in the usage.

**Comment by knaeckeKami** [comment]
That may very well be it.
After a short research, I found [this article](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html#access-bucket-intro).
and found
> Buckets created in Regions launched after March 20, 2019 are not reachable via the https://bucket.s3.amazonaws.com naming scheme. 

Did you create your bucket after that?

I will probably have to upload a new version of this package.

**Comment by ketanm** [comment]
This bucket was created much earlier. There are some differences in the canonical structure being created when compared to boto3. I will look into it further and figure out what is going on.

Also, since I am on Dart 2.5, the package as such does not compile. To test the package, I had to manually set everything to any.

**Comment by knaeckeKami** [comment]
Good to know, I should have tested it on earlier dart versions, too.

you can try testing the path-styled version of this library, currently on a separate branch:
pubspec.yaml:
```
  flutter_aws_s3_client:
    git:
      url: git://github.com/diagnosia/flutter_aws_s3_client.git
      ref: path_style_url
```

For my use cases, both url-styles work. 
I suspect it has something to do with the host, otherwise I don't know why a certificate error would apply.

Do you have any characters in your bucket-ID, that need to be url-encoded? (i.e. does pasting your bucket-id in https://www.url-encode-decode.com/ and pressing "encode url" output the original bucket id?)

**Comment by ketanm** [comment]
I do have bucket name as 'com.company.bucket' style approach

**Comment by knaeckeKami** [comment]
Ah, that explains the issue. It's a restriction of TLS-Wildcard-Certificates. It should work when using the path-style url.

> When you use virtual hosted–style buckets with Secure Sockets Layer (SSL), the SSL wildcard certificate only matches buckets that don't contain periods. To work around this, use HTTP or write your own certificate verification logic. We recommend that you do not use periods (".") in bucket names when using virtual hosted–style buckets. 

https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html

**Comment by ketanm** [comment]
Thanks for digging. I suspected but thought it should not be an issue 😁.

I will revisit again using path.

Ketan

**Comment by knaeckeKami** [comment]
Did you try if it works with path-styled URLs? (branch path_style_url of this repository)

If so, please give me feedback whether it worked for you, then I can clean the code up a little, merge it and release a new version.

**Comment by ketanm** [comment]
I tried and it worked fine. I have a few suggestions, I can create new issue if needed. Let me know.

My recommendations are:

1. Change example to refer to a local repository. I have a cloned branch and working on it but due to a locked version, it will be harder to see changes. This will enable to test changes and submit PR requests.

```
diff --git a/example/pubspec.yaml b/example/pubspec.yaml
index bfb41e5..fef5bef 100644
--- a/example/pubspec.yaml
+++ b/example/pubspec.yaml
@@ -19,7 +19,8 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
-  flutter_aws_s3_client: ^0.0.3
+  flutter_aws_s3_client:
+    path: ../
   # The following adds the Cupertino Icons font to your application.
   # Use with the CupertinoIcons class for iOS style icons.
   cupertino_icons: ^0.1.2
```

2. The package uses some versions which force the user to upgrade to latest Dart SDK. Currently, 2.5 would be used in many places, so forcing user to upgrade and be at the latest may not enable easy consumption. I did the following changes to make it work for me. (Note: Ignore package version as I copied my old file to test it out)
```
diff --git a/pubspec.yaml b/pubspec.yaml
index b79fb97..841095a 100644
--- a/pubspec.yaml
+++ b/pubspec.yaml
@@ -1,6 +1,6 @@
 name: flutter_aws_s3_client
 description: A simple, unofficial AWS S3 client in dart. Supports downloading objects and listing objects in a bucket.
-version: 0.4.1
+version: 0.3.0
 author: Diagnosia <support@diagnosia.com>
 homepage: https://github.com/diagnosia/flutter_aws_s3_client
 environment:
@@ -8,14 +8,14 @@ environment:

 dependencies:
   amazon_cognito_identity_dart: ^0.0.22
-  built_value: ^7.0.0
-  built_collection: ^4.2.0
-  xml2json : ^4.0.0
+  built_value: any
+  built_collection: any
+  xml2json : ^3.0.3
   http: ^0.12.0+2
 dev_dependencies:
-  built_value_generator: ^7.0.0
-  build_runner: ^1.7.2
-  json_serializable: ^3.2.2
+  built_value_generator: any
+  build_runner: any
+  json_serializable: any
   flutter_test:
     sdk: flutter
```

In pubspec.locked I have following after using `any`, it has moved to dev from main, not idea but would like package to work with 2.5 SDK.
built_value_generator - direct dev / "6.8.2"
build_runner - direct dev / "1.7.1"
json_serializable - direc dev / "3.2.3"

Regards
Ketan

**Comment by knaeckeKami** [comment]
Thanks for your improvements. I'll check them out.

**Comment by ketanm** [comment]
Hi @knaeckeKami, if you plan to merge branch on master then I can close this issue at that time or now as needed. Let me know if you need me to do any testing.

**Comment by knaeckeKami** [comment]
Hi @ketanm 
I will dedicate some time for that soon, hopefully this week.
I would let this issue open until I released a new version.

I hope in the meantime, it's okay for you to use the unreleased branch.

**Comment by knaeckeKami** [comment]
Hi @ketanm 

I added some changes. I changed the dependency of the example to a relative path.
I relaxed the version constraints. I did not set them so any but allowed a larger version range. Please check if that causes problems for you.

**Comment by knaeckeKami** [comment]
I released version 0.5.0, which should fix this issue.
If you still run into problems, please let me know.
I am closing this issue now.

**Comment by ketanm** [comment]
Thanks. I am busy with release. Will report issue if needed
